src/dataflow/src/render/reduce.rs

```diff
@@ -642,6 +735,80 @@ where
     negated_output.negate().concat(&input).consolidate()
 }
 
+fn build_hierarchical_monotonic<G>(
+    collection: Collection<G, (Row, Row)>,
+    aggrs: Vec<(usize, AggregateExpr)>,
+    prepend_key: bool,
+) -> Arrangement<G, Row>
+where
+    G: Scope,
+    G::Timestamp: Lattice,
+{
+    let aggr_funcs: Vec<_> = aggrs.iter().cloned().map(|(_, expr)| expr.func).collect();
```

Comment:
build_hierarchical and build_hierarchical_monotonic duplicate this whole block of code